src/targeting/build-page-targeting.ts

```diff
@@ -0,0 +1,151 @@
+import type { Participations } from '@guardian/ab-core';
+import { cmp } from '@guardian/consent-management-platform';
+import type { ConsentState } from '@guardian/consent-management-platform/dist/types';
+import type { CountryCode } from '@guardian/libs';
+import { getCookie, isString } from '@guardian/libs';
+import type { False, True } from '../types';
+import type { ContentTargeting } from './content';
+import { getContentTargeting } from './content';
+import type { AdManagerGroup, Frequency } from './personalised';
+import { getPersonalisedTargeting } from './personalised';
+import type { SessionTargeting } from './session';
+import { getSessionTargeting } from './session';
+import type { SharedTargeting } from './shared';
+import { getSharedTargeting } from './shared';
+import type { ViewportTargeting } from './viewport';
+import { getViewportTargeting } from './viewport';
+
+type PartialWithNulls<T> = { [P in keyof T]?: T[P] | null };
+
+type TrueOrFalse = True | False;
+
+type PageTargeting = PartialWithNulls<
+	{
+		ab: string[];
+		af: 't'; // Ad Free
+		amtgrp: AdManagerGroup;
+		at: string; // Ad Test
+		bp: 'mobile' | 'tablet' | 'desktop'; // BreakPoint
+		cc: CountryCode; // Country Code
+		cmp_interaction: string;
+		consent_tcfv2: string;
+		dcre: TrueOrFalse; // DotCom-Rendering Eligible
+		fr: Frequency; // FRequency
+		inskin: TrueOrFalse; // InSkin
+		pa: TrueOrFalse; // Personalised Ads consent
+		permutive: string[]; // predefined segment values
+		pv: string; // ophan Page View id
+		rdp: string;
+		ref: string; // REFerrer
+		rp: 'dotcom-rendering' | 'dotcom-platform'; // Rendering Platform
+		s: string; // site Section
+		sens: TrueOrFalse; // SenSitive
+		si: TrueOrFalse; // Signed In
+		skinsize: 'l' | 's';
+		urlkw: string[]; // URL KeyWords
+		vl: string; // Video Length
+
+		// And more
+		[_: string]: string | string[];
+	} & SharedTargeting
+>;
+
+const filterEmptyValues = (pageTargets: Record<string, unknown>) => {
+	const filtered: Record<string, string | string[]> = {};
+	for (const key in pageTargets) {
+		const value = pageTargets[key];
+		if (isString(value)) {
+			filtered[key] = value;
+		} else if (
+			Array.isArray(value) &&
+			value.length > 0 &&
+			value.every(isString)
+		) {
+			filtered[key] = value;
+		}
+	}
+	return filtered;
+};
+
+const buildPageTargeting = (
+	consentState: ConsentState,
+	adFree: boolean,
+	countryCode: CountryCode,
+	clientSideParticipations: Participations,
+): PageTargeting => {
+	if (
+		!window.guardian.config?.page ||
+		!window.guardian.config.ophan ||
+		window.guardian.config.isDotcomRendering === undefined
+	) {
+		return {};
+	}
+
+	const { page } = window.guardian.config;
+
+	const { isDotcomRendering } = window.guardian.config;
+
+	const adFreeTargeting: PageTargeting = adFree ? { af: 't' } : {};
```

Comment:
I assume this is just a straight migration but whilst we're here could we apply a more narrow type to this? Maybe

```suggestion
	const adFreeTargeting: { af?: 't' } = adFree ? { af: 't' } : {};
```

Reply:
Good shout, have changed!